modules/dag-l0/src/main/scala/io/constellationnetwork/dag/l0/infrastructure/rewards/GlobalDelegatedRewardsDistributor.scala

```diff
@@ -105,12 +67,60 @@ object DelegatedRewardsDistributor {
           facilitators,
           delegatedStakeDiffs,
           partitionedRecords,
-          Amount(NonNegLong.unsafeFrom(0L))
+          List.empty,
+          BigDecimal(0L),
+          BigDecimal(0L)
         )
       case TimeTrigger =>
-        calculateTotalRewardsToMint(epochProgress).flatMap(
-          applyDistribution(lastSnapshotContext, epochProgress, facilitators, delegatedStakeDiffs, partitionedRecords, _)
-        )
+        for {
+          emitFromFunction <- calculateTotalRewardsToMint(epochProgress)
+          pctConfig <- delegatedRewardsConfig.percentDistribution
+            .get(environment)
+            .pure[F]
+            .flatMap(Async[F].fromOption(_, new RuntimeException(s"Could not retrieve program distribution config for env: $environment")))
+          (reservedRewards, facilitatorRewardPool, delegatorRewardPool) <- calculateEmissionDistribution(emitFromFunction, pctConfig)
+          result <- applyDistribution(
+            lastSnapshotContext,
+            epochProgress,
+            facilitators,
+            delegatedStakeDiffs,
+            partitionedRecords,
+            reservedRewards,
+            facilitatorRewardPool,
+            delegatorRewardPool
+          )
+        } yield result
+    }
+
+    private def calculateEmissionDistribution(
+      totalRewards: Amount,
+      pctConfig: ProgramsDistributionConfig
+    ): F[(List[(Address, BigDecimal)], BigDecimal, BigDecimal)] = Async[F].delay {
+      val reservedAddressWeights = pctConfig.weights.toList.map { case (addr, pct) => addr -> pct.toBigDecimal }
+
+      val reservedWeight = reservedAddressWeights.map(_._2).sum
+      val validatorsWeight = pctConfig.validatorsWeight.toBigDecimal
+      val delegatorsWeight = pctConfig.delegatorsWeight.toBigDecimal
+      val totalWeight = reservedWeight + validatorsWeight + delegatorsWeight
+
+      val delegatorFlatInflationPercentage = delegatedRewardsConfig.flatInflationRate.toBigDecimal
+
+      val reservedAddressRewards =
+        if (totalWeight <= 0) List.empty
```

Comment:
I think this logic should be updated to return zero if the totalRewards is zero not the totalWeight. I will change this